### PR TITLE
WIP: Use `tg_file_id` for handling `file_id` stuff.

### PR DESCRIPTION
**Pull request by luckydonald**
Work in progress for #1613.

### CODE REVIEW

File: telethon/utils.py [hunk 1]
```diff
@@ -1101,104 +1103,85 @@ def resolve_bot_file_id(file_id):
...
-    data = _rle_decode(_decode_telegram_base64(file_id))
-    if not data:
+    if not file_id:
         return None
-
-    # This isn't officially documented anywhere, but
-    # we assume the last byte is some kind of "version".
-    data, version = data[:-1], data[-1]
-    if version not in (2, 4):
+    # end if
```
Comment (Lonami) [inline telethon/utils.py:1108]: Telethon's codebase doesn't really use this style of comments, so I think it's better to remove them.

File: telethon/utils.py [hunk 2]
```diff
@@ -1101,104 +1103,85 @@ def resolve_bot_file_id(file_id):
...
-        else:
-            return None
+    elif not isinstance(file, PhotoFileId):  # e.g. WebLocationFileId
+        return None
+    # end if
 
-        if not (1 <= dc_id <= 5):
-            return None
+    # okey, so we're dealing with a photo.
+    assert isinstance(file, PhotoFileId)
```
Comment (Lonami) [inline telethon/utils.py:1155]: I think I'd rather have this `return None` if it's not a photo:

```suggestion
    if not isinstance(file, PhotoFileId):
        return None
```

**Review by Lonami — CHANGES_REQUESTED** [review]
Looks alright except for some style nits (as with most PRs; style would be less of an issue if I used [`black`](https://pypi.org/project/black/)…).

File: telethon/utils.py [hunk 3]
```diff
@@ -1101,104 +1103,85 @@ def resolve_bot_file_id(file_id):
...
 
-        # Thumbnails (small) always have ID 0; otherwise size 'x'
-        photo_size = 's' if media_id or access_hash else 'x'
-        return types.Photo(
-            id=media_id,
-            access_hash=access_hash,
-            file_reference=b'',
-            date=None,
-            sizes=[types.PhotoSize(
+    foto_size = file.photosize
```
Comment (Lonami) [inline telethon/utils.py:1158]: Let's make this a bit more readable and give this an actual English name (and change the uses below too):

```python
    photo_type = 's' if file.id or file.access_hash else 'x' 
    photo_size = file.photosize
    if isinstance(photo_size, (
        PhotoFileId.PhotosizeSourceLegacy,
        PhotoFileId.PhotosizeSourceThumbnail,
        PhotoFileId.PhotosizeSourceDialogPhotoSmall,
        PhotoFileId.PhotosizeSourceDialogPhotoBig,
        PhotoFileId.PhotosizeSourceStickersetThumbnail,
    )):
```

File: setup.py [hunk 4]
```diff
@@ -232,7 +232,15 @@ def main(argv):
             packages=find_packages(exclude=[
                 'telethon_*', 'tests*'
             ]),
-            install_requires=['pyaes', 'rsa'],
+
+            # List run-time dependencies here. These will be installed by pip when your
+            # project is installed. For an analysis of "install_requires" vs pip's
+            # requirements files see:
```
Comment (Lonami) [inline setup.py:238]: The link is fine, but I think this is unnecessary. We're not here to teach others how to write a `setup.py` file.

**Comment by Lonami** [comment]
Closing since I think we settled we're not going to add this? Licensing and stuff IIRC.

**Comment by luckydonald** [comment]
:/

Thanks @danog.

**Comment by Lonami** [comment]
To be fair it shouldn't be hard to reverse engineer the format with a Bot API bot, sending and receiving all media types, and printing the results both in Telethon and the bot API to compare how things match up. Doing so would enable us to produce a new clean implementation. I personally haven't had the time or interest to do so, but if you do, feel free to open another PR.

I also haven't looked super hard at licensing compatibility or what it entails. It might actually be fine to merge this, but IANAL.

**Pull request closed without merging**